### PR TITLE
fix: allow layer flushes more often

### PR DESCRIPTION
**Pull request by koivunej**
As seen with the pgvector 0.7.0 index builds, we can receive large batches of images, leading to very large L0 layers in the range of 1GB. These large layers are produced because we are only able to roll the layer after we have witnessed two different Lsns in a single `DataDirModification::commit`. As the single Lsn batches of images can span over multiple `DataDirModification` lifespans, we will rarely get to write two different Lsns in a single `put_batch` currently.

The solution is to remember the TimelineWriterState instead of eagerly forgetting it until we really open the next layer or someone else flushes (while holding the write_guard). 

Additional changes are test fixes to avoid "initdb image layer optimization" or ignoring initdb layers for assertion.

Cc: #7197 because small `checkpoint_distance` will now trigger the "initdb image layer optimization"

### CODE REVIEW

**Comment by koivunej** [comment]
> * `test_actually_duplicated_l1`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7927/9321375051/index.html#suites/e789ce33839b63ec2f66607cbb480703/2c61fd24d191c23b), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7927/9321375051/index.html#suites/e789ce33839b63ec2f66607cbb480703/175f58368bf84261)

This test failure is rather interesting; it seems with low enough checkpoint_distance this new version will roll the layer before the wal is ingested, and so is able to perform the "initdb image layer optimization", or #5863.

**Comment by koivunej** [comment]
> * `test_fast_growing_tenant[absolute]`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7927/9321375051/index.html#suites/0e58fb04d9998963e98e45fe1880af7d/85fdbf2c73a88da9), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7927/9321375051/index.html#suites/0e58fb04d9998963e98e45fe1880af7d/ce51f33a996ad2bc)

This is an instance of the same problem, but now it does seem like the initdb is split into multiple image layers!

**Comment by koivunej** [comment]
> * `test_tenant_s3_restore`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7927/9368812016/index.html#suites/ab3c169995af0ceb1e11789889f9fa8d/6092533d4421886a), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7927/9368812016/index.html#suites/ab3c169995af0ceb1e11789889f9fa8d/4c6160f8931728fd)

This was most complicated so far, commented here: https://github.com/neondatabase/neon/issues/7197#issuecomment-2147915819

File: test_runner/regress/test_sharding.py [hunk 1]
```diff
@@ -712,6 +715,14 @@ def test_sharding_ingest_layer_sizes(
     tenant_id = env.initial_tenant
     timeline_id = env.initial_timeline
 
+    initial_layers_per_shard = {}
```
Comment (koivunej) [inline test_runner/regress/test_sharding.py:718]: should comment why we are doing this (initdb creates many "Small" image layers).

**Comment by koivunej** [comment]
The build status is red due ECR images becoming outdated. I assume we don't merge this without discussion, so not re-running the related jobs right now.

Discussion: <https://neondb.slack.com/archives/C059ZC138NR/p1717583177302419>

**Comment by koivunej** [comment]
Manual testing with gin index build: https://github.com/neondatabase/cloud/issues/13759#issuecomment-2152938601

**Comment by koivunej** [comment]
Testing to see how long did the gin test on main take, on the branch earlier it was ~16s per each of three tests.

File: pageserver/src/tenant/timeline.rs [hunk 2]
```diff
@@ -5691,6 +5689,14 @@ impl<'a> TimelineWriter<'a> {
             return OpenLayerAction::Open;
         };
 
+        if state.cached_last_freeze_at < self.tl.last_freeze_at.load() {
+            assert!(
+                state.open_layer.end_lsn.get().is_some(),
+                "our open_layer must be outdated"
+            );
```
Comment (koivunej) [inline pageserver/src/tenant/timeline.rs:5696]: Is this assert correct? It doesn't fire in manual testing or with the test suite.

We always first write down the end_lsn then increment the `last_freeze_at`.

Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:5696]: The write_guard can be dragged down to here `LayerManager::try_freeze_whatever`, but the changes are quite scary looking, will do that on a follow-up.

Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:5696]: Until then, this assertion is valid, with that refactoring, we can remove this as it's dead code... Or make it a `#[cfg(feature = "testing")]` assertion, because then it will be clear that whenever we freeze we will clear this state out.

Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:5696]: We support a case where we don't have an open layer but we still we advance the `Timeline::last_freeze_at` to `last_record_lsn + 1`.

However it doesn't matter for the assertion, because we only have TimelineWriterState and get to this assertion when there is an inmemory layer which could be frozen.

File: test_runner/regress/test_ingestion_layer_size.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        initial_tenant_conf={
+            "checkpoint_distance": f"{checkpoint_distance}",
+            "compaction_period": "0s",
+            "gc_period": "0s",
+            "compaction_threshold": "255",
+            "image_creation_threshold": "99999",
+        }
+    )
+
+    # build a 391MB gin index which exhibits the same behaviour as the pgvector with the two phase build
```
Comment (koivunej) [inline test_runner/regress/test_ingestion_layer_size.py:27]:
```suggestion
    # build a larger than checkpoint_distance gin index
    # building a gin index exhibits the same behaviour as does pgvector with the two phase build
```

Comment (koivunej) [reply test_runner/regress/test_ingestion_layer_size.py:27]: With double the rows it was a 391MB but I must've forgotten to drop this. with 64MB checkpoint_distance it is enough to build half the size. Could we we could half it once more, but it's already 16s to build.

Comment (koivunej) [reply test_runner/regress/test_ingestion_layer_size.py]: [75496dd](https://github.com/neondatabase/neon/pull/7927/commits/75496dd7ddddd25dfe1fff07ac1e8a7291226ab0)

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:5696]: I created #7993 to separate the refactorings.

File: test_runner/regress/test_disk_usage_eviction.py [hunk 4]
```diff
@@ -668,11 +655,11 @@ def test_fast_growing_tenant(neon_env_builder: NeonEnvBuilder, pg_bin: PgBin, or
         finish_tenant_creation(env, tenant_id, timeline_id, min_expected_layers)
 
     tenant_layers = count_layers_per_tenant(env.pageserver, map(lambda x: x[0], timelines))
-    (total_on_disk, _, _) = poor_mans_du(env, map(lambda x: x[0], timelines), env.pageserver, False)
+    (total_on_disk, _, _) = poor_mans_du(env, map(lambda x: x[0], timelines), env.pageserver, True)
 
     # cut 10 percent
```
Comment (VladLazar) [inline test_runner/regress/test_disk_usage_eviction.py:660]: nit: comment is stale now, although i'm not sure why this change was required

**Review by VladLazar — APPROVED** [review]
Nice test! Looks good to me - a few nits only

**Comment by koivunej** [comment]
> #### Failures on Postgres 15
> 
> * `test_location_conf_churn[2]`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7927/9446789776/index.html#suites/231aea51ea102d83d3dae52232693454/8b01cefe22fc0f04)

Vectored get validation panic while shutting down.

File: pageserver/src/tenant/timeline.rs [hunk 5]
```diff
@@ -3652,7 +3653,10 @@ impl Timeline {
         let _write_guard = if write_lock_held {
             None
         } else {
-            Some(self.write_lock.lock().await)
+            let mut g = self.write_lock.lock().await;
+            // remove the reference to an open layer
+            g.take();
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3658]: I'd rather not since the follow-up refactoring clears that up? At least moves it to happen only in a single place... Well, it doesn't hurt, I'll find a place.

File: test_runner/regress/test_disk_usage_eviction.py (hunk 4 above)
Comment (koivunej) [reply test_runner/regress/test_disk_usage_eviction.py:660]: the false -> true logs the layers which we ended up with before assertions, leaving something useful to the log about what was the overall state before the assertion failed.

good spotting with the obsolete comment.

File: pageserver/src/tenant/timeline.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs:3658]: Feel free to skip, but I think it would still be useful for future refactors even if with the follow up pr. Don't feel strongly at all though.

File: pageserver/src/tenant/timeline.rs [hunk 6]
```diff
@@ -5691,6 +5689,14 @@ impl<'a> TimelineWriter<'a> {
             return OpenLayerAction::Open;
         };
 
+        if state.cached_last_freeze_at < self.tl.last_freeze_at.load() {
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:5697]: 7f23c2f35

File: pageserver/src/tenant/timeline.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3660]: 5db078092
